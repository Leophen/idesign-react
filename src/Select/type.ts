import { DropdownOption } from "../Dropdown/type";

export interface SelectProps {
  /**
   * 类名
   */
  className?: string;
  /**
   * 内容
   */
  children?: React.ReactNode;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 选择器宽度
   * @default 100%
   */
  width?:React.CSSProperties["width"];
  /**
   * 固定选中值（受控）
   */
  value?: string | number | Array<string | number>;
  /**
   * 默认选中值（非受控）
   * @default []
   */
  defaultValue?: string | number | Array<string | number>;
  /**
   * 占位符
   * @default 请选择
   */
  placeholder?: string;
  /**
   * 下拉操作项
   * @default []
   */
  options?: Array<DropdownOption>;
  /**
   * 选择器尺寸
   * @default medium
   */
  size?: "small" | "medium" | "large";
  /**
   * 是否可一键清空
   * @default true
   */
  clearable?: boolean;
  /**
   * 选择框前置图标名
   */
  prefixIcon?: string;
  /**
   * 选择框后置图标名
   */
  suffixIcon?: string;
  /**
   * 选择框前置图标类名
   */
  prefixIconClass?: string;
  /**
   * 选择框后置图标类名
   */
  suffixIconClass?: string;
  /**
   * 级联子层级展开方向
   * @default right
   */
  cascaderDirection?: 'left' | 'right';
  /**
   * 是否可多选
   * @default false
   */
  multiple?: boolean;
  /**
   * 是否禁用选择器
   * @default false
   */
  disabled?: boolean;
  /**
   * 选中值变化时触发
   */
  onChange?: (value: string | number | Array<string | number>) => void;
}

export interface SelectItemProps {
  /**
   * 内容
   */
  children?: React.ReactNode;
}
