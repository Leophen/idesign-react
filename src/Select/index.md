---
nav:
  title: 组件
  path: /components
group:
  title: 表单组件
  order: 3
order: 5
---

# Select 选择器

下拉选择器。

## 基本用法

可通过包裹 `<Select.Item>` 或传入 `options` 参数来实现选择器，可将 `clearable` 设为 `false` 关闭一键清空功能。

```tsx
import React, { useState } from 'react'
import { Select } from 'idesign-react'

const App = () => {
  const [value, setValue] = useState('item1')

  const options = [
    {
      content: '选项一',
      value: 'item1'
    },
    {
      content: '选项二',
      value: 'item2'
    },
    {
      content: '选项三',
      value: 'item3'
    }
  ]

  const onChange = (val) => {
    console.log(val)
    setValue(val)
  }

  return (
    <div className="idesign-demo-block-row">
      <Select value={value} onChange={onChange}>
        <Select.Item value="item1">选项一</Select.Item>
        <Select.Item value="item2">选项二</Select.Item>
        <Select.Item value="item3">选项三</Select.Item>
      </Select>
      <Select
        value={value}
        options={options}
        clearable={false}
        onChange={onChange}
      />
    </div>
  )
}

export default App
```

## 不同尺寸

可通过 `size` 属性来指定不同尺寸的选择器，默认为 `medium`。

```tsx
import React, { useState } from 'react'
import { Select } from 'idesign-react'

const App = () => {
  const [value, setValue] = useState('item1')

  const options = [
    {
      content: '选项一',
      value: 'item1'
    },
    {
      content: '选项二',
      value: 'item2'
    },
    {
      content: '选项三',
      value: 'item3'
    }
  ]

  const onChange = (val) => {
    console.log(val)
    setValue(val)
  }

  return (
    <div className="idesign-demo-block-row">
      <Select
        value={value}
        options={options}
        size="small"
        onChange={onChange}
      />
      <Select
        value={value}
        options={options}
        size="medium"
        onChange={onChange}
      />
      <Select
        value={value}
        options={options}
        size="large"
        onChange={onChange}
      />
    </div>
  )
}

export default App
```

## 禁用选择器

可通过 `disabled` 属性来禁用选择器。

```tsx
import React, { useState } from 'react'
import { Select } from 'idesign-react'

const App = () => {
  const [value, setValue] = useState('item1')

  const options = [
    {
      content: '选项一',
      value: 'item1'
    },
    {
      content: '选项二',
      value: 'item2'
    },
    {
      content: '选项三',
      value: 'item3'
    }
  ]

  const onChange = (val) => {
    console.log(val)
    setValue(val)
  }

  return (
    <div className="idesign-demo-block-row">
      <Select
        value={value}
        options={options}
        disabled={true}
        onChange={onChange}
      />
    </div>
  )
}

export default App
```

## 单项禁用

可在 `<Select.Item>` 使用 `disabled` 属性来控制单项禁用，也可在 `<Select>` 的 `options` 属性中设置 `disabled` 控制单项禁用。

```tsx
import React, { useState } from 'react'
import { Select } from 'idesign-react'

const App = () => {
  const [value, setValue] = useState('item1')

  const options = [
    {
      content: '选项一',
      value: 'item1'
    },
    {
      content: '选项二',
      value: 'item2',
      disabled: true
    },
    {
      content: '选项三',
      value: 'item3'
    }
  ]

  const onChange = (val) => {
    console.log(val)
    setValue(val)
  }

  return (
    <div className="idesign-demo-block-row">
      <Select value={value} onChange={onChange}>
        <Select.Item value="item1">选项一</Select.Item>
        <Select.Item value="item2" disabled>选项二</Select.Item>
        <Select.Item value="item3">选项三</Select.Item>
      </Select>
      <Select
        value={value}
        options={options}
        onChange={onChange}
      />
    </div>
  )
}

export default App
```

## 单项带分割线

可在 `<Select.Item>` 使用 `divider` 属性来控制单项禁用，也可在 `<Select>` 的 `divider` 属性中设置 `disabled` 控制单项禁用。

```tsx
import React, { useState } from 'react'
import { Select } from 'idesign-react'

const App = () => {
  const [value, setValue] = useState('item1')

  const options = [
    {
      content: '选项一',
      value: 'item1'
    },
    {
      content: '选项二',
      value: 'item2',
      divider: true
    },
    {
      content: '选项三',
      value: 'item3'
    }
  ]

  const onChange = (val) => {
    console.log(val)
    setValue(val)
  }

  return (
    <div className="idesign-demo-block-row">
      <Select value={value} onChange={onChange}>
        <Select.Item value="item1">选项一</Select.Item>
        <Select.Item value="item2" divider>选项二</Select.Item>
        <Select.Item value="item3">选项三</Select.Item>
      </Select>
      <Select
        value={value}
        options={options}
        onChange={onChange}
      />
    </div>
  )
}

export default App
```

## 单项带分组标题

可在 `<Select.Item>` 使用 `title` 属性来控制单项禁用，也可在 `<Select>` 的 `title` 属性中设置 `disabled` 控制单项禁用。

```tsx
import React, { useState } from 'react'
import { Select } from 'idesign-react'

const App = () => {
  const [value, setValue] = useState('item1')

  const options = [
    {
      content: '选项一',
      value: 'item1',
      title: '分组 1'
    },
    {
      content: '选项二',
      value: 'item2',
      divider: true
    },
    {
      content: '选项三',
      value: 'item3',
      title: '分组 2'
    }
  ]

  const onChange = (val) => {
    console.log(val)
    setValue(val)
  }

  return (
    <div className="idesign-demo-block-row">
      <Select value={value} onChange={onChange}>
        <Select.Item value="item1" title="分组 1">选项一</Select.Item>
        <Select.Item value="item2" divider>选项二</Select.Item>
        <Select.Item value="item3" title="分组 2">选项三</Select.Item>
      </Select>
      <Select
        value={value}
        options={options}
        onChange={onChange}
      />
    </div>
  )
}

export default App
```

## 带前后缀图标

可通过 `<Select>` 的 `prefixIcon` 和 `suffixIcon` 属性设置选择框前后缀图标。

```tsx
import React, { useState } from 'react'
import { Select } from 'idesign-react'

const App = () => {
  const [value, setValue] = useState('item1')

  const options = [
    {
      content: '选项一',
      value: 'item1'
    },
    {
      content: '选项二',
      value: 'item2'
    },
    {
      content: '选项三',
      value: 'item3'
    }
  ]

  const onChange = (val) => {
    console.log(val)
    setValue(val)
  }

  return (
    <div className="idesign-demo-block-row">
      <Select
        value={value}
        options={options}
        prefixIcon="Search"
        onChange={onChange}
      />
      <Select
        value={value}
        options={options}
        suffixIcon="Search"
        onChange={onChange}
      />
    </div>
  )
}

export default App
```

## 多选选择器

可通过 `multiple` 属性设为多选选择器。

```tsx
import React, { useState } from 'react'
import { Select } from 'idesign-react'

const App = () => {
  const [value, setValue] = useState([])

  const options = [
    {
      content: '选项一',
      value: 'item1'
    },
    {
      content: '选项二',
      value: 'item2'
    },
    {
      content: '选项三',
      value: 'item3'
    }
  ]

  const onChange = (val) => {
    console.log(val)
    setValue(val)
  }

  return (
    <div className="idesign-demo-block-row">
      <Select
        value={value}
        options={options}
        multiple={true}
        onChange={onChange}
      />
      <Select
        value={value}
        options={options}
        multiple={true}
        onChange={onChange}
      />
    </div>
  )
}

export default App
```

## 级联选择器

可在 `options` 属性中为单项添加 `children` 属性指定级联菜单，另外可通过 `cascaderDirection` 来控制展开方向。

```tsx
import React, { useState } from 'react'
import { Select } from 'idesign-react'

const App = () => {
  const [value, setValue] = useState([])

  const options = [
    {
      content: '操作 1',
      value: 1
    },
    {
      content: '操作 2',
      value: 2,
      children: [
        {
          content: '操作 2-1',
          value: 2.1
        },
        {
          content: '操作 2-2',
          value: 2.2,
          children: [
            {
              content: '操作 2-2-1',
              value: 2.21
            },
            {
              content: '操作 2-2-2',
              value: 2.22
            },
            {
              content: '操作 2-2-3',
              value: 2.23
            }
          ]
        },
        {
          content: '操作 2-3',
          value: 2.3
        }
      ]
    },
    {
      content: '操作 3',
      value: 3
    }
  ]

  const onChange = (val) => {
    console.log(val)
    setValue(val)
  }

  return (
    <div className="idesign-demo-block-row">
      <Select
        placeholder="级联选择器"
        value={value}
        options={options}
        onChange={onChange}
      />
      <Select
        placeholder="级联多选选择器"
        value={value}
        options={options}
        multiple={true}
        onChange={onChange}
      />
      <Select
        placeholder="向左展开的级联选择器"
        value={value}
        options={options}
        cascaderDirection="left"
        onChange={onChange}
      />
      <Select
        placeholder="向左展开的级联多选选择器"
        value={value}
        options={options}
        cascaderDirection="left"
        multiple={true}
        onChange={onChange}
      />
    </div>
  )
}

export default App
```

## Select API

<API hideTitle />

## SelectItem API

| 属性 | 说明 | 类型 | 默认值 |
| --- | --- | --- | --- |
| content | 下拉操作项内容 | `ReactNode` | `--` |
| value | 下拉操作项唯一标识 | `string〡number` | `--` |
| disabled | 单项是否禁用 | `boolean` | `false` |
| divider | 是否显示操作项之后的分隔线 | `boolean` | `false` |
| title | 该下拉项上方组标题 | `string` | `--` |
| onClick | 点击时触发 | `(dropdownItem: DropdownOption, event: React.MouseEvent) => void` | `--` |
