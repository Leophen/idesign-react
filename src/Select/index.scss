.i-select {
  width: 100%;
  .i-popup__reference {
    width: 100%;
    & > .i-input {
      height: auto;
      .i-input__inner {
        &::selection {
          background: transparent;
        }
      }
    }
  }
  &[data-size='small'] {
    width: 160px;
  }
  &[data-size='large'] {
    width: 240px;
  }
  .i-input__hide-clear {
    & > .icon-TipCloseFill {
      display: none !important;
    }
  }
  .i-input {
    & > .icon-TipCloseFill {
      cursor: pointer;
    }
    & > .icon-ArrowDown {
      pointer-events: none;
    }
  }
  .i-tag--close-btn .i-icon {
    cursor: pointer;
  }
}

.i-select-arrow__show {
  transform: rotate(180deg);
}

.i-select__multiple-wrap {
  display: flex;
  flex-wrap: wrap;
  gap: 6px;
  align-items: center;
  width: 100%;
  margin: 6px 0;
  overflow: hidden;
  .i-tag {
    border-radius: 2px;
  }
}
