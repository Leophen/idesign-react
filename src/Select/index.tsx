import React, { useEffect, useState, useRef } from 'react';
import classNames from 'classnames';
import _ from 'lodash';
import './index.scss';
import Dropdown from '../Dropdown';
import { DropdownOption } from '../Dropdown'
import Input from '../Input';
import Tag from '../Tag';

export interface SelectProps {
  /**
   * 类名
   */
  className?: string;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 选择器宽度
   * @default 100%
   */
  width?: string | number;
  /**
   * 选中值
   */
  value?: string | number | Array<string | number>;
  /**
   * 占位符
   * @default 请选择
   */
  placeholder?: string;
  /**
   * 下拉操作项
   * @default []
   */
  options?: Array<DropdownOption>;
  /**
   * 是否可多选
   * @default false
   */
  multiple?: boolean;
  /**
   * 选中值变化时触发
   */
  onChange?: (value: string | number | Array<string | number>) => void;
}


const SelectItem: React.FC = (props) => {
  const { children } = props;
  return (
    <>{children}</>
  )
}

const Select: React.FC<SelectProps> & { Item: React.ElementType } = (props) => {
  const {
    children,
    className,
    style,
    width,
    value = [],
    placeholder = '请选择',
    options = [],
    multiple = false,
    onChange = () => { },
    ...others
  } = props;

  // 转换下拉数据及更新下拉框宽度
  const selectNode = useRef<HTMLDivElement>(null)
  const [innerOptions, setInnerOptions] = useState(options)
  const [dropdownWidth, setDropdownWidth] = useState(0)
  useEffect(() => {
    // Select.Item 模式 -> 更新下拉数据
    if (children) {
      let selectData: any = []
      React.Children.map(children, (child) => {
        if (!React.isValidElement(child)) {
          return null;
        }
        let currentActive = false
        if (innerValue === child.props.value) {
          currentActive = true
        }
        selectData.push({
          content: child.props.children,
          value: child.props.value,
          active: child.props.active || currentActive,
        })
      })
      setInnerOptions(selectData)
    }
    // 更新下拉宽度
    const currentWidth = selectNode.current?.getBoundingClientRect().width || 0
    setDropdownWidth(currentWidth)
  }, [])

  // 更新下拉数据
  const [updateKey, setUpdateKey] = useState(0)
  const [innerValue, setInnerValue] = useState(value)

  const handleSelect = (val: string | number | Array<string | number>) => {
    setInnerValue(val)
    setUpdateKey(new Date().getTime())
    onChange?.(val)
  }

  const getItemContent = (val: string | number | Array<string | number>) => {
    let content = ''
    innerOptions.map(item => {
      if (item.value === val && item.content) {
        content = item.content.toString()
      }
    })
    return content
  }

  // 选择器显示隐藏操作
  const [dropdownShow, setDropdownShow] = useState(false)
  const handleTrigger = (visible: boolean) => {
    setDropdownShow(visible)
  }

  // 删除多选项
  const handleDelItem = (e: React.MouseEvent, val: string | number) => {
    e.stopPropagation()
    if (Array.isArray(innerValue)) {
      const curInnerValue = innerValue
      let delIndex = 0
      curInnerValue.map((item, index) => item === val && (delIndex === index))
      curInnerValue.splice(delIndex, 1)
      setUpdateKey(new Date().getTime())
      setInnerValue(curInnerValue)
    }
  }

  return (
    <div
      ref={selectNode}
      className={classNames(
        'i-select',
        className
      )}
      style={{ ...(style || {}), ...{ width } }}
      {...others}
    >
      <Dropdown
        width={dropdownWidth}
        options={innerOptions}
        onClick={handleSelect}
        onTrigger={handleTrigger}
        value={innerValue}
        selected={true}
        multiple={multiple}
      >
        <Input
          value={!multiple ? getItemContent(innerValue) : undefined}
          placeholder={(!multiple || (Array.isArray(innerValue) && innerValue.length) === 0) ? placeholder : ''}
          readonly
          suffixIcon="ArrowDown"
          suffixIconClass={dropdownShow ? "i-select-arrow__show" : ""}
        >
          {multiple && Array.isArray(innerValue) && innerValue.length > 0 && (
            <div className='i-select__multiple-wrap'>
              {
                innerValue.map(val => (
                  <Tag
                    theme="dark"
                    size="small"
                    onClose={(e: React.MouseEvent) => handleDelItem(e, val)}
                    key={val}
                  >
                    {getItemContent(val)}
                  </Tag>
                ))
              }
            </div>
          )}
        </Input>
      </Dropdown>
    </div>
  );
};

Select.Item = SelectItem;

SelectItem.displayName = 'SelectItem';
Select.displayName = 'Select';

export default Select;
