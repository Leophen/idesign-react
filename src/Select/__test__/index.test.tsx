import { useState } from 'react'
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Select from '../index';

describe('Select 组件测试', () => {
  it('create', () => {
    const { asFragment } = render(
      <Select>
        <Select.Item value="item1">选项一</Select.Item>
        <Select.Item value="item2">选项二</Select.Item>
        <Select.Item value="item3">选项三</Select.Item>
      </Select>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('onChange', async () => {
    const SingleSelect = () => {
      const [value, setValue] = useState('item2');
      const onChange = (value) => {
        setValue(value);
      };
      return (
        <Select value={value} onChange={onChange}>
          <Select.Item value="item1">选项一</Select.Item>
          <Select.Item value="item2">选项二</Select.Item>
          <Select.Item value="item3">选项三</Select.Item>
        </Select>
      );
    };
    const { container, getByText } = render(<SingleSelect />);

    // 未点击 input 前，popup 不出现
    expect(document.querySelector('.i-popup')).toBeNull();

    // 鼠标点击 input，popup 出现
    const user = userEvent.setup()
    await user.click(container.querySelector('.i-input'))

    const popupElement = document.querySelector('.i-popup')
    expect(popupElement).not.toBeNull();
    expect(popupElement).toHaveTextContent('选项一');
    expect(popupElement).toHaveTextContent('选项二');
    expect(popupElement).toHaveTextContent('选项三');

    // 点击'选项一'选项，input 展示该选项，且 popup 消失
    await user.click(getByText('选项一'))

    expect(document.querySelector('.i-input__inner')).toHaveValue('选项一');
    expect(document.querySelector('.i-popup')).toBeInTheDocument();
  });

  it('size', () => {
    const { container } = render(
      <Select size="small">
        <Select.Item value="item1">选项一</Select.Item>
        <Select.Item value="item2">选项二</Select.Item>
        <Select.Item value="item3">选项三</Select.Item>
      </Select>
    );
    expect(container.querySelector('.i-input--size-small')).toBeInTheDocument();
  });

  it('disabled', async () => {
    const { container } = render(
      <Select disabled={true}>
        <Select.Item value="item1">选项一</Select.Item>
        <Select.Item value="item2">选项二</Select.Item>
        <Select.Item value="item3">选项三</Select.Item>
      </Select>
    );
    expect(container.querySelector('.i-input-is-disabled')).toBeInTheDocument();

    const user = userEvent.setup()
    await user.click(container.querySelector('.i-input'))

    expect(document.querySelector('.i-popup')).toBeNull();
  });

  it('single disabled', async () => {
    const { container, getByText } = render(
      <Select>
        <Select.Item value="item1" disabled>选项一</Select.Item>
        <Select.Item value="item2">选项二</Select.Item>
        <Select.Item value="item3">选项三</Select.Item>
      </Select>
    );
    expect(document.querySelector('.i-popup')).toBeNull();

    const user = userEvent.setup()
    await user.click(container.querySelector('.i-input'))

    expect(document.querySelector('.i-popup')).not.toBeNull();

    await user.click(getByText('选项一'))

    expect(document.querySelector('.i-popup')).not.toBeNull();
  });

  it('title', async () => {
    const { container, getByText } = render(
      <Select>
        <Select.Item value="item1" title="分组 1">选项一</Select.Item>
        <Select.Item value="item2">选项二</Select.Item>
        <Select.Item value="item3">选项三</Select.Item>
      </Select>
    );

    const user = userEvent.setup()
    await user.click(container.querySelector('.i-input'))

    expect(getByText('分组 1')).toBeInTheDocument();
  });
});



