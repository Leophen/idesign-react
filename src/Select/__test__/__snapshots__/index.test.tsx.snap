// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select 组件测试 create 1`] = `
<DocumentFragment>
  <div
    class="i-select"
  >
    <div
      class="i-popup__reference"
    >
      <div
        class="i-input i-input-is-readonly"
      >
        <input
          class="i-input__inner"
          placeholder="请选择"
          readonly=""
          value=""
        />
        <i
          class="i-icon icon-ArrowDown i-input-suffix-icon"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
